lib/services/access_request.go

```diff
@@ -1075,26 +1091,73 @@ func (m *RequestValidator) Validate(ctx context.Context, req types.AccessRequest
 
 		now := m.clock.Now().UTC()
 
-		// Calculate expiration time of the Access Request (how long it
+		// Calculate the expiration time of the Access Request (how long it
 		// will await approval).
 		ttl, err := m.requestTTL(ctx, identity, req)
 		if err != nil {
 			return trace.Wrap(err)
 		}
 		req.SetExpiry(now.Add(ttl))
 
-		// Calculate expiration time of the elevated certificate that will
-		// be issued if the Access Request is approved.
-		ttl, err = m.sessionTTL(ctx, identity, req)
+		maxDuration, err := m.calculateMaxAccessDuration(req)
 		if err != nil {
 			return trace.Wrap(err)
 		}
-		req.SetAccessExpiry(now.Add(ttl))
+		// If the maxDuration flag is set, use it instead of the session TTL.
+		if maxDuration > 0 {
+			ttl = maxDuration
+		} else {
+			// Calculate the expiration time of the elevated certificate that will
+			// be issued if the Access Request is approved.
+			ttl, err = m.sessionTTL(ctx, identity, req)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+
+		accessTTL := now.Add(ttl)
+		req.SetAccessExpiry(accessTTL)
 	}
 
 	return nil
 }
 
+// calculateMaxAccessDuration calculates the maximum time for the access request.
+// The max duration time is the minimum of the max_duration time set on the request
+// and the max_duration time set on the request role.
+func (m *RequestValidator) calculateMaxAccessDuration(req types.AccessRequest) (time.Duration, error) {
+	// Check if the maxDuration time is set.
+	maxDurationTime := req.GetMaxDuration()
+	if maxDurationTime.IsZero() {
+		return 0, nil
+	}
+
+	maxDuration := maxDurationTime.Sub(req.GetCreationTime())
+	if maxDuration < 0 {
+		return 0, trace.BadParameter("invalid maxDuration: must be greater than creation time")
+	}
+
+	if maxDuration > maxAccessDuration {
+		return 0, trace.BadParameter("max_duration must be less or equal 7 days")
+	}
+
+	minAdjDuration := maxDuration
+	// Adjust the expiration time if the max_duration value is set on the request role.
+	for _, tms := range m.MaxDurationMatchers {
+		for _, matcher := range tms.Matchers {
+			for _, roleName := range req.GetRoles() {
+				if matcher.Match(roleName) {
+					if tms.MaxDuration < minAdjDuration {
+						minAdjDuration = tms.MaxDuration
+					}
+				}
+			}
+		}
+	}
```

Comment:
I think this for-loop should be tweaked.  I see two potential issues:

-   As I understand this, if no `MaxDurationMatchers` entries matcha given role, then `minAdjDuration` is not affected, but thats not really what we want.  If there is no `MaxDuration` defined for a role, it means you're not allowed to request that role with an extended duration.  I.e. the "default" max duration for a role is zero, and should be applied as such.
- If a given role has both a `max_duration: 2d` and a `max_duration: 4d` that match it, this loop will apply the `2d` limit, but I think we want the opposite.  For a given role, we should select the maximum value... then we should constrain the `minAdjDuration` by the smallest of the per-role maximums.

Double-check my logic, but I think we can achieve the above like so:

```suggestion
	for _, roleName := range req.GetRoles() {
		var maxDurationForRole time.Duration
		for _, tms := range m.MaxDurationMatchers {
			for _, matcher := range tms.Matchers {	
				if matcher.Match(roleName) {
					if tms.MaxDuration > maxDurationForRole {
						maxDurationForRole = tms.MaxDuration
					}
				}
			}
		}
		
		if maxDurationForRole < minAdjDuration {
			minAdjDuration = maxDurationForRole
		}
	}
```